builder.md

```diff
@@ -1,57 +1,180 @@
 
 - `scheme-name` - obtained from the scheme file
 - `scheme-author` - obtained from the scheme file
 - `scheme-slug` - obtained from the scheme filename, as described above
+- `scheme-kind` - either "light" or "dark", calculated from the `base00` color <!-- TODO: This is a candidate for inclusion, let me know your thoughts -->
+
+- `base00-hex` to `base0F-hex` - obtained from the scheme file e.g "7cafc2"
 - `base00-hex-r` to `base0F-hex-r` - built from the hex value in the scheme file e.g "7c"
 - `base00-hex-g` to `base0F-hex-g` - built from the hex value in the scheme file e.g "af"
 - `base00-hex-b` to `base0F-hex-b` - built from the hex value in the scheme file e.g "c2"
+- `base00-hex-bgr` to `base0F-hex-bgr` - built from a reversed version of all the hex values e.g "c2af7c"
+
 - `base00-rgb-r` to `base0F-rgb-r` - converted from the hex value in the scheme file e.g "124"
 - `base00-rgb-g` to `base0F-rgb-g` - converted from the hex value in the scheme file e.g "175"
 - `base00-rgb-b` to `base0F-rgb-b` - converted from the hex value in the scheme file e.g "194"
 - `base00-dec-r` to `base0F-dec-r` - converted from the rgb value in the scheme file e.g "0.87..."
 - `base00-dec-g` to `base0F-dec-g` - converted from the rgb value in the scheme file e.g "0.50..."
 - `base00-dec-b` to `base0F-dec-b` - converted from the rgb value in the scheme file e.g "0.21..."
 
-Builders should also provide the following variables for convenience:
+## 3. Interface
 
-- `base00-hex` to `base0F-hex` - obtained from the scheme file e.g "7cafc2"
-- `base00-hex-bgr` to `base0F-hex-bgr` - built from a reversed version of all the hex values e.g "c2af7c"
+All base16 reference builders MUST provide a single feature: building a
+template using 1 or more schemes.
+
+This section (`3`) is not relevant to extended tooling, which will have its own
+interface fitting their intended usecases, altough they MAY expose the
+reference CLI interface (through a `build` subcommand, for example) and/or a
+library, if desired.
+
+It is REQUIRED that this functionality is exposed by one (or both) of the following:
+
+### 3.1. CLI
+
+The binary name SHOULD contain `base16`, but is otherwise left as a choice to
+the author.
+
+To be fully compliant, a builder CLI interface MUST NOT include any other
+feature, option, argument, or deviance from the expected interface and
+behaviour of the program.
+
+<!-- TODO: For convenience, we make a manpage and set of tests available. All
+compliant builders MUST fully conform to these two. -->
+
+The CLI interface MUST work without relying on any network connection.
+
+#### 3.1.1 CLI Arguments and options
+
+```bash
+base16 <TEMPLATES-DIRECTORY> <SCHEME-FILE> ...
+```
+
+`TEMPLATES-DIRECTORY` being a directory containing a `config.yaml`, as well as at
+least one `.mustache` template. These are usually named `templates` on all
+base16 template repositories.
+
+`SCHEME-FILE` being one or more `.yaml` scheme file. The main source for these
```

Comment:
Additionally/alternatively, we should specify that schemes should be loaded (by default) from the base16-schemes repo.

Reply:
Sorry, yet another comment here - one thing I like about the base16-builder-go is the ability to embed a default set of schemes, but allow overriding it if the user wants to clone the schemes repo. Is that something we want to support?

Reply:
> I'd like to allow for passing a directory of schemes as well to allow building all schemes at once.

I think traversing the directory (should it be recursive or not? is there a max depth? etc) should be handled by the shell. That's why `SCHEME-FILE` accepts multiple files. This allows, for example:

```bash
base16 template schemes/*.yaml
```

```bash
base16 template $(find . -name *.yaml)
```

Also, not implementing the traversal logic allows easier implementation: all files that should be opened are known by the program without listing files:
- `TEMPLATES-DIRECTORY/config.yaml`
- Templates files specified by the previous yaml
- All files specified by `SCHEME-FILE`

Reply:
> Additionally/alternatively, we should specify that schemes should be loaded (by default) from the base16-schemes repo.

> Sorry, yet another comment here - one thing I like about the base16-builder-go is the ability to embed a default set of schemes, but allow overriding it if the user wants to clone the schemes repo. Is that something we want to support?

Hmm, I'd rather not couple the builders with the repo. This would also require all of them to have git as dependencies.

I think that explicit and agnostic is better than implicit and coupled (at least when we're talking about plumbing), so I'd argue that asking the scripter to `git clone` first would be preferrable. Would that make sense?

Reply:
1. If we move to passed arguments rather than a single directory (I probably wouldn't recurse, at least for now), we need to add back the error when scheme slugs overlap.
2. At least right now, everything is fairly coupled and I personally think that's fine. It lets us make things convenient because we control the spec, schemes, and some builder implementations. If it was coupled, we could say that builders should embed the schemes or have a method of pulling them from github (but provide a method of overriding it, maybe a command line flag). This is already an extremely opinionated change (the CLI spec), so maybe it's worth using some of those opinions to make things more convenient to use.

Reply:
I'm also just now realizing you could use the base16-builder on the base16-builder repos to programatically generate files with the color schemes for including in the builder 🤣 .

Reply:
> one thing I like about the base16-builder-go is the ability to embed a default set of schemes

I just considered this and then quickly ruled it out because it's a maintenance nightmare for a tiny bit of convenience... unless we were going to auto-publishing a new builder every time a theme was added... 

I did submodule the schemes so if someone does a git checkout they get them, but even that might annoy me in the future since it's just one more thing that needs to be automated or manually maintained vs just included `git clone` in the directions.

Reply:
I actually just updated this - I recently added a command line flag allowing the user to run in "online mode" which pulls the schemes directly from github (using the tarball endpoint - it was surprisingly easy) and I default to that in the provided Github action, but falls back to the "embedded" schemes.

Reply:
What would we think of having a --schemes-dir argument and allowing multiple arguments for which templates to process?

Reply:
I really dislike making plumbing tools interact with git or somehow maintain state, so I'd rather be Unixy and make the _reference_ builders just implement the template spec. That is, just build a template + a number of schemes into a number of themes/configs/outputs.

Reply:
Also on the unixyness part, I think iterating over files on a directory is also not very good when we're talking about plumbing tools. There are much better tools available for that (regular shell globbing, find, etc) that can easily be used to get whichever schemes the caller might prefer.

I don't have a strong opinion about building just 1 or multiple templates, as long as we can do something like:
```bash
git clone https://github.com/base16-project/base16-schemes
git clone https://github.com/base16-project/base16-shell
git clone https://github.com/base16-project/base16-vim

base16 --templates base16-{shell,vim}/templates --schemes base16-schemes/*
```